### ISSUE TITLE
Poor Inference Quality After Converting HF Model to NeMo [LLama 3.1]

### ISSUE THREAD

**Issue by Harsha0056** [NVIDIA/NeMo#14668]
After finetuning, the results were quite poor and the model hallucinated a lot. To debug, I decided to test the base model with general questions. However, the base model’s responses were also very bad (sometimes even worse than the finetuned one).

If the HF base model converted to NeMo itself produces such degraded responses, I don’t think I can expect the finetuned model to work well.

**Question:** 1. Is there a way to fix this issue or improve the base → NeMo conversion so that the responses at least match the HF base model performance?
2. Or do i need to tweak any parameters ?

**Additional Info:**

Below are comparisons between NeMo-converted base model vs original HF

1. Above result  -  Nemo base converted back  to HF response:
2. Below result -    Base model response

<img width="1479" height="612" alt="Image" src="https://github.com/user-attachments/assets/9a61df80-059f-4aea-9ffa-6a415ee3afe0" />

**Used Inference use code  for Nemo base :**

<img width="662" height="869" alt="Image" src="https://github.com/user-attachments/assets/be542790-9ce9-4e8f-810e-acdbdaf650c0" />

**Hf Base converted to nemo  :**
<img width="1439" height="154" alt="Image" src="https://github.com/user-attachments/assets/06c9f346-8952-4636-becb-274b98903d49" />

** Same Model  for 0.3 temp**

<img width="1466" height="281" alt="Image" src="https://github.com/user-attachments/assets/bcacda2e-b0e4-4f72-b421-1b71bc2542f3" />

**Edit : 
Response of a finetuned model [ on pudmedq dataset ] which was converted back  to hf  format**

<img width="1408" height="777" alt="Image" src="https://github.com/user-attachments/assets/ab313cb8-808d-4d55-9ced-9946970fed73" />

**Comment by mehraakash** [comment]
Thanks for sharing the details. To help narrow this down, could you let us know which NeMo container (image and tag/version) you’re using for the conversion and finetuning?

**Comment by Harsha0056** [comment]
Hi @mehraakash ,Thanks for your response. I tested with both the 25.04 and 25.07 containers the quality was same, and we’re using NeMo 2.0 for both conversions and fine-tuning. These are the docs we followed.

Edit : 
Link :  https://docs.nvidia.com/nemo-framework/user-guide/latest/sft_peft/peft_nemo2.html
https://docs.nvidia.com/nemo-framework/user-guide/latest/nemo-2.0/features/hf-integration.html

**Comment by Harsha0056** [comment]
@mehraakash 
We deployed the NeMo base model after converting it back to Hugging Face format inside a container. The responses were noticeably better in this setup, but we ran into two issues: we can’t pass any chat template, and the model doesn’t know when to stop [eos is missing]—it keeps repeating the same answer.
 In contrast, with a NIM-based model  , we’re able to use chat templates without this problem also able to pass chat templates and model does know when to stop .

<img width="1919" height="510" alt="Image" src="https://github.com/user-attachments/assets/cb05cd42-e9c9-4128-9217-2646dad6e7ca" />

<img width="1919" height="758" alt="Image" src="https://github.com/user-attachments/assets/7466ae03-ca01-4c65-8486-79691cf88617" />

adding my colleague in this loop @Shrii-WorkspaceNSX

**Comment by Harsha0056** [comment]
@mehraakash 
We deployed a fine-tuned model trained on the PubMedQ dataset (validation loss: 0.39), converted it back to Hugging Face format, and ran it in a container. However, even on general questions (not related to PubMedQ), it tends to hallucinate.
Note: While we are able to pass chat templates, the overall performance remains poor.

<img width="1352" height="736" alt="Image" src="https://github.com/user-attachments/assets/fc8e62c9-7cdf-4365-a0df-de04c3f91f8e" />

<img width="1839" height="884" alt="Image" src="https://github.com/user-attachments/assets/41507b67-680b-4a3b-b5ca-d9666aa59b13" />

**Tested with other base model:** 
**Quality with llama 3.2 3B**
**Question** : Talk about atlantis

<img width="1194" height="311" alt="Image" src="https://github.com/user-attachments/assets/2a4138bb-f2a9-466b-be19-67d406757e7e" />

**Quality with qwen 2 1.5B**
**Question** : Talk about atlantis
<img width="1197" height="615" alt="Image" src="https://github.com/user-attachments/assets/cde590b4-1eed-428a-a593-debf3c23eb4b" />

**Quality with Nemotron llama3.1**
**Question** : Talk about atlantis

<img width="1209" height="240" alt="Image" src="https://github.com/user-attachments/assets/8add7f84-3681-4fc6-bab3-039b9eaa41cc" />

**Comment by pzelasko** [comment]
Closing this issue as "won't fix".
The following collections have been moved to separate repos in https://github.com/NVIDIA-NeMo organization: avlm, llm, multimodal, multimodal-autoregressive, vlm, speechlm, diffusion
Please move the discussion to the related repo if the issue persists.
